collect_app/src/main/java/org/odk/collect/android/audio/AudioRecordingControllerFragment.java

```diff
@@ -61,49 +64,64 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
 
     @Override
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
-        audioRecorder.getCurrentSession().observe(getViewLifecycleOwner(), session -> {
-            if (session == null) {
-                binding.getRoot().setVisibility(GONE);
-            } else if (session.getFailedToStart() != null) {
-                binding.getRoot().setVisibility(GONE);
-                showIfNotShowing(AudioRecordingErrorDialogFragment.class, getParentFragmentManager());
-            } else if (session.getFile() == null) {
-                binding.getRoot().setVisibility(VISIBLE);
-
-                binding.timeCode.setText(LengthFormatterKt.formatLength(session.getDuration()));
-                binding.waveform.addAmplitude(session.getAmplitude());
-
-                if (session.getPaused()) {
-                    binding.pauseRecording.setIcon(ContextCompat.getDrawable(getContext(), R.drawable.ic_baseline_mic_24));
-                    binding.pauseRecording.setContentDescription(getString(R.string.resume_recording));
-                    binding.pauseRecording.setOnClickListener(v -> audioRecorder.resume());
-
-                    binding.recordingIcon.setImageDrawable(ContextCompat.getDrawable(getContext(), R.drawable.ic_pause_24dp));
-                } else {
-                    binding.pauseRecording.setIcon(ContextCompat.getDrawable(getContext(), R.drawable.ic_pause_24dp));
-                    binding.pauseRecording.setContentDescription(getString(R.string.pause_recording));
-                    binding.pauseRecording.setOnClickListener(v -> {
-                        audioRecorder.pause();
-                        formEntryViewModel.logFormEvent(AnalyticsEvents.AUDIO_RECORDING_PAUSE);
-                    });
-
-                    binding.recordingIcon.setImageDrawable(ContextCompat.getDrawable(getContext(), R.drawable.ic_baseline_mic_24));
-                }
-
-                // Pause not available before API 24
-                if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.N) {
-                    binding.pauseRecording.setVisibility(GONE);
-                }
-
-                if (backgroundAudioViewModel.isBackgroundRecording()) {
-                    binding.pauseRecording.setVisibility(GONE);
-                    binding.stopRecording.setVisibility(GONE);
-                }
-            } else {
-                binding.getRoot().setVisibility(GONE);
-            }
+        zip3(formEntryViewModel.hasBackgroundRecording(), backgroundAudioViewModel.isBackgroundRecordingEnabled(), audioRecorder.getCurrentSession()).observe(getViewLifecycleOwner(), triple -> {
```

Comment:
I looked at removing the zip here and seperating it out to different observations but I ended up in trouble as I had to make some of the same decisions in each observation (the `null` session case needs to reach out to `hasBackgroundRecording` vice versa).

I also tried using a "React-esque" approach of setting fields on the fragment for each observation and then updating, but I ended up in trouble as the first observation would fire before the next and so with unset fields being using to make decisions.

Reply:
Thanks for looking into it. Let's go with this.

Reply:
Let's pair on this when we can as I think it's probably a nice test of how we're going to use `LiveData`.